### PR TITLE
PE-4352: Track Custom JSON Metadata and Custom GQL Tags in the DB

### PR DESCRIPTION
**Pull request by matibat**
--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/0qsd4cr8tocd0

### CODE REVIEW

File: lib/entities/entity.dart [hunk 1]
```diff
@@ -71,6 +74,124 @@ abstract class Entity {
...
 
+abstract class EntityWithCustomMetadata extends Entity {
+  // The custom JSON Metadata sub-JSON.
+  /// These are the keys in the JSON Metadata, excluding the reserved ones.
+  @JsonKey(includeFromJson: false, includeToJson: false)
+  Map<String, dynamic>? customJsonMetadata = {};
+
+  // The reserved JSON Metadata keys.
+  @JsonKey(includeFromJson: false, includeToJson: false)
+  abstract final List<String> reservedJsonMetadataKeys;
```
Comment (thiagocarvalhodev) [inline lib/entities/entity.dart:85]: Why do we need the `abstract` keyword here?

File: lib/entities/entity.dart [hunk 2]
```diff
@@ -71,6 +74,124 @@ abstract class Entity {
   void addEntityTagsToTransaction<T extends TransactionBase>(T tx);
 }
 
+abstract class EntityWithCustomMetadata extends Entity {
```
Comment (thiagocarvalhodev) [inline lib/entities/entity.dart:77]: In accordance with the Single Responsibility Principle, an Entity should ideally have a single responsibility to ensure clarity and maintainability.

I recognize that this pattern emerged based on prior implementations, but as we look ahead, we might consider using an Adapter or Builder pattern in future refactoring to ensure that each component only handles one specific responsability.

File: test/entities/custom_json_metadata_test.dart [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+      fileDataItem.addTag('banana', 'manzana');
+      folderDataItem.addTag('banana', 'manzana');
+      driveDataItem.addTag('banana', 'manzana');
+
+      fileData = Uint8List.fromList(utf8.encode(jsonEncode(fakeFileJson)));
+      folderData = Uint8List.fromList(utf8.encode(jsonEncode(fakeFolderJson)));
+      driveData = Uint8List.fromList(utf8.encode(jsonEncode(fakeDriveJson)));
+    });
+
+    // test('for FileEntity', () async {
```
Comment (thiagocarvalhodev) [inline test/entities/custom_json_metadata_test.dart:164]: Remove commented code

File: lib/models/file_entry.dart [hunk 4]
```diff
@@ -1,17 +1,30 @@
...
+      size: size,
+      lastModifiedDate: lastModifiedDate,
+      dataContentType: dataContentType,
+      pinnedDataOwnerAddress: pinnedDataOwnerAddress,
+    );
+    file.customJsonMetadata =
+        customJsonMetadata != null ? jsonDecode(customJsonMetadata!) : null;
+    file.customGqlTags = customGQLTags != null
+        ? (jsonDecode(customGQLTags!) as List<dynamic>)
+            .map((maybeTag) => Tag.fromJson(maybeTag))
```
Comment (thiagocarvalhodev) [inline lib/models/file_entry.dart:25]: Why `maybeTag`?

Comment (thiagocarvalhodev) [inline lib/models/file_entry.dart:25]: nit: Consider implementing a private method or a utils method for parsing a list of tags to a better readability

File: lib/entities/entity.dart [hunk 5]
```diff
@@ -71,6 +74,124 @@ abstract class Entity {
...
+    EntityTag.unixTime,
+    EntityTag.cipher,
+    EntityTag.cipherIv,
+    EntityTag.appName,
+    EntityTag.appVersion,
+    EntityTag.appPlatform,
+    EntityTag.input,
+    EntityTag.contract,
+    'Bundle-Format',
+    'Bundle-Version',
```
Comment (thiagocarvalhodev) [inline lib/entities/entity.dart:131]: nit: Add these strings on the `EntityTag` class

File: lib/entities/drive_entity.dart [hunk 6]
```diff
@@ -22,6 +22,22 @@ class DriveEntity extends Entity {
...
+    ...EntityWithCustomMetadata.sharedReservedGqlTags,
+    EntityTag.drivePrivacy,
+    EntityTag.driveAuthMode,
+  ];
+
+  @override
+  @JsonKey(includeFromJson: false, includeToJson: false)
+  List<String> reservedJsonMetadataKeys = [
+    ...EntityWithCustomMetadata.sharedReservedJsonMetadataKeys,
+    'rootFolderId',
```
Comment (thiagocarvalhodev) [inline lib/entities/drive_entity.dart:38]: Add this string to `EntityTag` class

File: lib/entities/file_entity.dart [hunk 7]
```diff
@@ -34,6 +34,24 @@ class FileEntity extends Entity {
...
+  ];
+
+  @override
+  @JsonKey(includeFromJson: false, includeToJson: false)
+  List<String> reservedJsonMetadataKeys = [
+    ...EntityWithCustomMetadata.sharedReservedJsonMetadataKeys,
+    'size',
+    'lastModifiedDate',
+    'dataTxId',
+    'dataContentType',
```
Comment (thiagocarvalhodev) [inline lib/entities/file_entity.dart:52]: nit: Add these strings to `EntityTag` class, or if doesn't fit well, consider implementing a new class for Metadata keys

**Review by thiagocarvalhodev — COMMENTED** [review]
Left some nits and a few questions.

File: lib/entities/entity.dart (hunk 1 above)
Comment (thiagocarvalhodev) [reply lib/entities/entity.dart:85]: I liked it!

Comment (matibat) [reply lib/entities/entity.dart:85]: > Why do we need the abstract keyword here?

To force its implementations to declare its own reserved keys

File: test/entities/custom_json_metadata_test.dart (hunk 3 above)
Comment (matibat) [reply test/entities/custom_json_metadata_test.dart:164]: I will either un-comment that code or create a tech debt to do that later on. In the end, we need the unit tests for GQL

File: lib/entities/drive_entity.dart (hunk 6 above)
Comment (matibat) [reply lib/entities/drive_entity.dart:38]: Ahmm these are not GQL tags but JSON Metadata Fields

File: lib/entities/file_entity.dart (hunk 7 above)
Comment (matibat) [reply lib/entities/file_entity.dart:52]: Same here: these are not GQL tags but JSON Metadata Fields

File: lib/entities/entity.dart (hunk 1 above)
Comment (thiagocarvalhodev) [reply lib/entities/entity.dart:85]: I never used it, but it is very helpful! Thanks for bringing that, @matibat

File: lib/entities/entity.dart (hunk 2 above)
Comment (thiagocarvalhodev) [reply lib/entities/entity.dart:77]: Oh, no. We are good. Just a reminder for us when we refactor this class

File: test/entities/custom_json_metadata_test.dart (hunk 3 above)
Comment (thiagocarvalhodev) [reply test/entities/custom_json_metadata_test.dart:164]: I prefer a tech debt

**Review by thiagocarvalhodev — APPROVED** [review]
LGTM

**Pull request merged**